File: Book/Haskell/Chapter 3/BruteForceSearch/BruteForceSearch.hs
module BruteForceSearch where

-- Input: array A of n distinct integers.
-- Output: the number of inversions of A.
bruteForceSearch :: (Ord a) => [a] -> Integer
bruteForceSearch xs =

    outer xs 0

    -- Go through every element, x, of the list.
    where outer [] rsf = rsf
          outer (x:xs) rsf =
            outer xs (rsf + (inner xs 0))

            -- Check for inversions:
            --  Elements that are smaller than x but that appear
            --   later in xs.
            where inner [] rsf = rsf
                  inner (x':xs) rsf =
                    inner xs (if x > x' then (succ rsf)
                                        else rsf)

File: Java/src/ListsAndArrays/CountInversions/BruteForceSearch.java
package ListsAndArrays.CountInversions;

public class BruteForceSearch {

     /**
     * Brute-Force Search for Counting Inversions
     *
     * @param xs
     * Input: array A of n distinct integers.
     *
     * @return numInv
     * Output: the number of inversions of A.
     */
    static int bruteForceSearch(int[] xs) {

        int numInv = 0;
        int n = xs.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (xs[i] > xs [j]) {
                    numInv++;
                }
            }
        }

        return numInv;
    }
}
